### PR TITLE
impl PartialEq<Vec<B>> for &[A], &mut [A]

### PR DESCRIPTION
**Pull request by sollyucko**
https://github.com/rust-lang/rfcs/issues/2917

### CODE REVIEW

**Comment by marmeladema** [comment]
I am definitely not a compiler-team member and I don't want to overstep any boundaries here, so I apologize in advance if I do, just let me know. But should this kind of change requires some tests?

**Comment by shepmaster** [comment]
I would agree that some tests should be added, specifically those exercising the specific equality being added. Also, I’d have some `assert_eq` usages as this comes up as a common usage.

**Comment by kennytm** [comment]
@shepmaster ya know I can't start the required FCP for this insta-stable PR since I'm not in the libs team 🙃

**Comment by shepmaster** [comment]
> ya know

I appreciate that you think I know _anything_. What I know is that you are a thorough reviewer (like one who realized that this would be insta-stable and thus needs a RFC).

File: src/liballoc/tests/vec.rs [hunk 1]
```diff
@@ -1473,3 +1473,61 @@ fn vec_macro_repeating_null_raw_fat_pointer() {
         vtable: *mut (),
     }
 }
+
+#[test]
+fn partialeq() {
+    let vec1 = vec![1, 2];
+    let vec2 = vec![1, 2, 3];
+    let slice1 = &[1, 2];
```
Comment (shepmaster) [inline src/liballoc/tests/vec.rs:1481]: This is not a slice; it's a reference to an array.

**Review by shepmaster — DISMISSED** [review]
The "slice"s are not slices; that needs to be changed.

You don't have to follow this, but I would have written the tests to lean on the type checker more. Something like:

```rust
fn compare_vec_and_array(v: Vec<i32>, a: [i32; 3]) {
    let _ = v == a;
    let _ = a == v;

    assert_eq!(v, a);
    assert_eq!(a, v);
}

fn compare_vec_and_slice(v: Vec<i32>, s: &[i32]) { /* ... */ }

// Optionally make these take in `Vec<T>`?
```

I don't know if I would even *call* these functions; their existence proves that the traits exist. I *might* call them just to prove that they were compiled.

I also advocate for having smaller tests with better names. That way, when a test fails, you can tell what failed by the name, and it's easier to see *what* is being tested.

**Comment by sollyucko** [comment]
I'm still keeping the mega-test for thoroughness, but I can remove it if you'd prefer.

File: src/liballoc/tests/vec.rs (hunk 1 above)
Comment (sollyucko) [reply src/liballoc/tests/vec.rs]: Fixed! I now have a proper slice and a separate reference to an array.

**Comment by sollyucko** [comment]
Should I create `PartialEq` implementations for arrays with `Vec`s, or remove the tests for them?

**Comment by sollyucko** [comment]
Actually, probably best for symmetry.

File: src/liballoc/tests/vec.rs [hunk 2]
```diff
@@ -1473,3 +1474,118 @@ fn vec_macro_repeating_null_raw_fat_pointer() {
...
+}
+
+fn assert_eq_vec_and_ref_array_3<T: PartialEq + Debug>(v: Vec<T>, a: &[T; 3]) {
+    let _ = v == a;
+    let _ = a == v;
+
+    assert_eq!(v, a);
+    assert_eq!(a, v);
+}
+
```
Comment (LukasKalbertodt) [inline src/liballoc/tests/vec.rs:1501]: There are tests missing for the `&mut [_]` case.

**Review by LukasKalbertodt — DISMISSED** [review]
Seems like a useful addition to me!

Could you add UI tests still for the array impls? Those should go somewhere in `src/tests/ui/const-generics/array-impls/`. You need to add to `alloc-traits-impls-length-32.rs` and `alloc-traits-no-impls-length-33.rs`. This ensures that we don't accidentally stabilize any impls for arrays longer than 32. To update the `.stderr`, you probably want to use the `--bless` flag for the test runner.

**Comment by LukasKalbertodt** [comment]
@shepmaster ya know I can't start the required FCP for this insta-stable PR since I'm still (deliberately) [excluded from the rfcbot list](https://github.com/rust-lang/team/blob/830961073b42de2c5350ac20c2ec48cb3ca8a603/teams/libs.toml#L31-L35) :upside_down_face:

.
.
.

:smile: 
I will reassign myself once the PR is ready!

**Comment by shepmaster** [comment]
> @shepmaster ya know I can't start the required FCP

You people assuming I know anything even when presented *with extreme evidence to the contrary* are sweet.

**Comment by sollyucko** [comment]
Re-review?

File: src/liballoc/vec.rs [hunk 3]
```diff
@@ -2340,15 +2340,20 @@ macro_rules! __impl_slice_eq1 {
...
+__impl_slice_eq1! { [] &mut [A], Vec<B>, }
 __impl_slice_eq1! { [] Cow<'_, [A]>, &[B], A: Clone }
 __impl_slice_eq1! { [] Cow<'_, [A]>, &mut [B], A: Clone }
 __impl_slice_eq1! { [] Cow<'_, [A]>, Vec<B>, A: Clone }
 __impl_slice_eq1! { [const N: usize] Vec<A>, [B; N], [B; N]: LengthAtMost32 }
 __impl_slice_eq1! { [const N: usize] Vec<A>, &[B; N], [B; N]: LengthAtMost32 }
+__impl_slice_eq1! { [const N: usize] Vec<A>, &mut [B; N], [B; N]: LengthAtMost32 }
+__impl_slice_eq1! { [const N: usize] [A; N], Vec<B>, [A; N]: LengthAtMost32 }
+__impl_slice_eq1! { [const N: usize] &[A; N], Vec<B>, [A; N]: LengthAtMost32 }
+__impl_slice_eq1! { [const N: usize] &mut [A; N], Vec<B>, [A; N]: LengthAtMost32 }
```
Comment (dtolnay) [inline src/liballoc/vec.rs:2353]: We should not do these until after https://github.com/rust-lang/rust/issues/61415 is resolved. If const generics ends up rolled back and these need to be reverted to an individual impl for each size, that is quite a bit of bloat (as indicated in the comment on the next line).

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: src/liballoc/vec.rs (hunk 3 above)
Comment (sollyucko) [reply src/liballoc/vec.rs:2353]: Is the issue with the size of the macro-generated code?

**Comment by sollyucko** [comment]
Yay, finally finished wrangling git! Final sequence of commands that *should* work next time:

```sh
git fetch --all
git checkout master
git reset --hard rust-lang/master
git rebase master origin/master
$EDITOR src/liballoc/tests/vec.rs
git add src/liballoc/tests/vec.rs
git rebase --continue
git checkout master
git merge --ff-only $THE_COMMIT_OUTPUT_BY_CHECKOUT
git push --force
```

File: src/liballoc/vec.rs (hunk 3 above)
Comment (dtolnay) [reply src/liballoc/vec.rs:2350]: Not specifically. Everything gets worse if the standard library were to contain 32 times as many impls. Rustdoc becomes a worse experience, diagnostics from the trait solver get worse, compile time gets worse, etc.

**Comment by dtolnay** [comment]
Waiting on https://github.com/rust-lang/rust/pull/71660#pullrequestreview-414220590.

File: src/liballoc/vec.rs [hunk 4]
```diff
@@ -2337,15 +2337,19 @@ macro_rules! __impl_slice_eq1 {
...
+__impl_slice_eq1! { [] &[A], Vec<B>, }
+__impl_slice_eq1! { [] &mut [A], Vec<B>, }
 __impl_slice_eq1! { [] Cow<'_, [A]>, &[B], A: Clone }
 __impl_slice_eq1! { [] Cow<'_, [A]>, &mut [B], A: Clone }
 __impl_slice_eq1! { [] Cow<'_, [A]>, Vec<B>, A: Clone }
 __impl_slice_eq1! { [const N: usize] Vec<A>, [B; N], [B; N]: LengthAtMost32 }
 __impl_slice_eq1! { [const N: usize] Vec<A>, &[B; N], [B; N]: LengthAtMost32 }
-
+__impl_slice_eq1! { [const N: usize] [A; N], Vec<B>, [A; N]: LengthAtMost32 }
+__impl_slice_eq1! { [const N: usize] &[A; N], Vec<B>, [A; N]: LengthAtMost32 }
```
Comment (dtolnay) [inline src/liballoc/vec.rs:2348]: Please remove these impls. (https://github.com/rust-lang/rust/pull/71660#discussion_r427125135)

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: src/liballoc/vec.rs (hunk 4 above)
Comment (sollyucko) [reply src/liballoc/vec.rs:2348]: Oops, will do.

**Comment by dtolnay** [comment]
@rust-lang/libs:
@rfcbot fcp merge

```rust
impl PartialEq<Vec<B>> for &[A] where A: PartialEq<B> {...}
impl PartialEq<Vec<B>> for &mut [A] where A: PartialEq<B> {...}
```

https://github.com/rust-lang/rfcs/issues/2917#issue-608440680 calls out that the reverse direction already exists: Vec\<A\> implements PartialEq\<&amp;\[B\]\> and PartialEq\<&amp;mut \[B\]\>. In addition the other PartialEq impls involving \[T\], &amp;\[T\], &amp;mut \[T\], and \[T; N\] are implemented symmetrically.

Probably the reason these two were left out was that it would have been disallowed by the orphan rule prior to Rust 1.41:

```console
error[E0210]: type parameter `A` must be used as the type parameter for some local type (e.g., `MyStruct<A>`)
    --> src/liballoc/vec.rs:2260:1
     |
2260 | impl<A, B> PartialEq<Vec<B>> for &[A] where A: PartialEq<B> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `A` must be used as the type parameter for some local type
     |
     = note: only traits defined in the current crate can be implemented for a type parameter
```

**Comment by jplatte** [comment]
This has the milestone set to 1.45, but it now can't go in before 1.46, right?

**Comment by mbrubeck** [comment]
As mentioned in https://github.com/rust-lang/rfcs/issues/2917#issuecomment-633243843 I think the following impls should also be included:

```rust
__impl_slice_eq1! { [] Vec<A>, [B], }
__impl_slice_eq1! { [] [A], Vec<B>, }
```

**Comment by dtolnay** [comment]
I reverted 73194c527519565a686e8e79d9f81ed8374a1d07 because those impls were not included in the FCP proposal. There can be a separate FCP in a different PR.

**Review by dtolnay — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
included in rollup (accidentally two of them :P ) 

@bors retry

**Pull request merged**